src/3d/symbols/qgspoint3dsymbol_p.cpp

```diff
@@ -0,0 +1,381 @@
+    geometry = g;
+  }
+  else if ( shape == "torus" )
+  {
+    float radius = shapeProperties["radius"].toFloat();
+    float minorRadius = shapeProperties["minorRadius"].toFloat();
+    Qt3DExtras::QTorusGeometry *g = new Qt3DExtras::QTorusGeometry;
+    g->setRadius( radius ? radius : 10 );
+    g->setMinorRadius( minorRadius ? minorRadius : 5 );
+    geometry = g;
+  }
+  else if ( shape == "plane" )
+  {
+    float size = shapeProperties["size"].toFloat();
+    Qt3DExtras::QPlaneGeometry *g = new Qt3DExtras::QPlaneGeometry;
+    g->setWidth( size ? size : 10 );
+    g->setHeight( size ? size : 10 );
+    geometry = g;
+  }
+#if QT_VERSION >= 0x050900
+  else if ( shape == "extrudedText" )
+  {
+    float depth = shapeProperties["depth"].toFloat();
+    QString text = shapeProperties["text"].toString();
+    Qt3DExtras::QExtrudedTextGeometry *g = new Qt3DExtras::QExtrudedTextGeometry;
+    g->setDepth( depth ? depth : 1 );
+    g->setText( text );
+    geometry = g;
+  }
+#endif
+  else  // shape == "cylinder" or anything else
+  {
+    float radius = shapeProperties["radius"].toFloat();
+    float length = shapeProperties["length"].toFloat();
+    Qt3DExtras::QCylinderGeometry *g = new Qt3DExtras::QCylinderGeometry;
+    //g->setRings(2);  // how many vertices vertically
+    //g->setSlices(8); // how many vertices on circumference
+    g->setRadius( radius ? radius : 10 );
+    g->setLength( length ? length : 10 );
+    geometry = g;
+  }
+
+  geometry->addAttribute( instanceDataAttribute );
+  geometry->setBoundingVolumePositionAttribute( instanceDataAttribute );
+
+  Qt3DRender::QGeometryRenderer *renderer = new Qt3DRender::QGeometryRenderer;
+  renderer->setGeometry( geometry );
+  renderer->setInstanceCount( count );
+
+  return renderer;
+}
+
+//* 3D MODEL RENDERING *//
+
+static Qt3DExtras::QPhongMaterial *phongMaterial( const QgsPoint3DSymbol &symbol )
+{
+  Qt3DExtras::QPhongMaterial *phong = new Qt3DExtras::QPhongMaterial;
+
+  phong->setAmbient( symbol.material().ambient() );
+  phong->setDiffuse( symbol.material().diffuse() );
+  phong->setSpecular( symbol.material().specular() );
+  phong->setShininess( symbol.material().shininess() );
+
+  return phong;
+}
+
+void QgsPoint3DSymbolModelEntityFactory::addEntitiesForSelectedPoints( const Qgs3DMapSettings &map, QgsVectorLayer *layer, const QgsPoint3DSymbol &symbol, QgsPoint3DSymbolEntity *parent )
+{
+  QgsFeatureRequest req;
+  req.setDestinationCrs( map.crs() );
+  req.setFilterFids( layer->selectedFeatureIds() );
+
+  addMeshEntities( map, layer, req, symbol, parent, true );
+}
+
+
+
+void QgsPoint3DSymbolModelEntityFactory::addEntitiesForNotSelectedPoints( const Qgs3DMapSettings &map, QgsVectorLayer *layer, const QgsPoint3DSymbol &symbol, QgsPoint3DSymbolEntity *parent )
+{
+  // build the feature request to select features
+  QgsFeatureRequest req;
+  req.setDestinationCrs( map.crs() );
+  QgsFeatureIds notSelected = layer->allFeatureIds();
```

Comment:
ouch - allFeatureIds is a very expensive call. Can we avoid this?

Reply:
I think so - but again something I would prefer to address at some point later...